src/AMAMM.sol

```diff
@@ -2,69 +2,200 @@
+        _updateLastUpdatedEpoch(id, _epoch);
+    }
+
+    function depositToken(PoolId id, address depositor, uint128 rent) internal returns (uint128) {
+        uint128 amount = uint128(rent * K(id));
+
+        if (_userBalance[depositor] >= amount) {
+            uint128 remainderAmount = uint128(_userBalance[depositor] - amount);
+            _pullBidToken(id, depositor, remainderAmount);
+            _userBalance[depositor] += remainderAmount;
+        } else {
+            _pullBidToken(id, depositor, amount);
+            _userBalance[depositor] += amount;
+        }
+
+        return amount;
+    }
+    /// @inheritdoc IAmAmm
+
+    function withdrawBalance(PoolId id, uint128 _amount) external virtual override returns (uint128) {
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        if (!_amAmmEnabled(id) || _userBalance[msgSender] < _amount) {
+            revert AmAmm__InvalidBid();
+        }
+
+        _userBalance[msgSender] -= _amount;
+        _pushBidToken(id, msgSender, _amount);
+
+        return _amount;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Internal helpers
+    /// -----------------------------------------------------------------------
+
+    /// @notice returns current epoch.
+    /// @param id pool id
+    /// @param _epoch current epoch
+    function _getDeposit(PoolId id, uint40 _epoch) public view returns (uint256) {
+        return uint256(poolEpochManager[id][_epoch].rent * uint256(K(id)));
     }
 
-    function afterSwap(
-        address /* sender **/,
-        PoolKey calldata key,
-        IPoolManager.SwapParams calldata params,
-        BalanceDelta delta,
-        bytes calldata /* hookData **/
-    ) external override poolManagerOnly returns (bytes4, int128) {
-        // fee will be in the unspecified token of the swap
-        bool specifiedTokenIs0 = (params.amountSpecified < 0 ==
-            params.zeroForOne);
-        (Currency feeCurrency, int128 swapAmount) = (specifiedTokenIs0)
-            ? (key.currency1, delta.amount1())
-            : (key.currency0, delta.amount0());
-        // if fee is on output, get the absolute output amount
-        if (swapAmount < 0) swapAmount = -swapAmount;
-
-        uint256 feeAmount = (uint128(swapAmount) * SWAP_FEE_BIPS) / TOTAL_BIPS;
-        // TODO: change address(this) to the ePoche manager address
-        poolManager.take(feeCurrency, address(this), feeAmount);
-
-        return (IHooks.afterSwap.selector, feeAmount.toInt128());
+    /// @notice returns current epoch.
+    /// @param id pool id
+    /// @param timestamp current timestamp
+    function _getEpoch(PoolId id, uint256 timestamp) public view returns (uint40) {
+        return uint40(timestamp / EPOCH_SIZE(id));
     }
+
+    function _getRefund(PoolId id, uint40 _epoch, uint40 _targetEpoch) internal view returns (uint256) {
+        return poolEpochManager[id][_epoch].rent * (_epoch + K(id) - _targetEpoch);
```

Comment:
It would be similar logic from the `getLastManager`

```suggestion
        if (_lastUpdatedEpoch[id] > targetEpoch) {
            if (_lastUpdatedEpoch[id] - targetEpoch <= K(id)) {
                return poolEpochManager[id][_lastUpdatedEpoch[id]].rent * (targetEpoch + K(id) - _lastUpdatedEpoch[id]);
                }
        } else {
            if (targetEpoch - _lastUpdatedEpoch[id] <= K(id)) {
                return poolEpochManager[id][_lastUpdatedEpoch[id]].rent * (_lastUpdatedEpoch[id] + K(id) - targetEpoch);
                }
        }
```

Reply:
I dont think this is needed because of where the _getRefund is being called. Its only being called inside those same if statement

 ```
       if (prevWinner.rent != 0) {
            if (prevWinner.rent < rent) {
                //Userp Top Bidder and only allow bidder to own N epochs instead of K epochs (N < K)
                _userBalance[prevWinner.bidder] += _getRefund(id, _lastUpdatedEpoch[id], _epoch); //Refund losing bidder
                ```